source/non_matching/fe_values.cc

```diff
@@ -0,0 +1,399 @@
+
+#include <deal.II/lac/block_vector.h>
+#include <deal.II/lac/la_parallel_block_vector.h>
+#include <deal.II/lac/la_parallel_vector.h>
+#include <deal.II/lac/la_vector.h>
+#include <deal.II/lac/petsc_block_vector.h>
+#include <deal.II/lac/petsc_vector.h>
+#include <deal.II/lac/trilinos_epetra_vector.h>
+#include <deal.II/lac/trilinos_parallel_block_vector.h>
+#include <deal.II/lac/trilinos_tpetra_vector.h>
+#include <deal.II/lac/trilinos_vector.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/non_matching/fe_values.h>
+
+#include <deal.II/numerics/fe_field_function.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace NonMatching
+{
+  RegionUpdateFlags::RegionUpdateFlags()
+    : inside(update_default)
+    , outside(update_default)
+    , surface(update_default)
+  {}
+
+
+  namespace internal
+  {
+    namespace FEValuesImplementation
+    {
+      template <int dim, class VectorType>
+      RefSpaceFEFieldFunction<dim, VectorType>::RefSpaceFEFieldFunction(
+        const DoFHandler<dim> &dof_handler,
+        const VectorType &     dof_values)
+        : dof_handler(&dof_handler)
+        , global_dof_values(&dof_values)
+      {
+        Assert(dof_handler.n_dofs() == dof_values.size(),
+               ExcDimensionMismatch(dof_handler.n_dofs(), dof_values.size()));
+      }
+
+
+
+      template <int dim, class VectorType>
+      void
+      RefSpaceFEFieldFunction<dim, VectorType>::set_active_cell(
+        const typename Triangulation<dim>::active_cell_iterator &cell)
+      {
+        Assert(
+          &cell->get_triangulation() == &dof_handler->get_triangulation(),
+          ExcMessage(
+            "The incoming cell must belong to the triangulation associated with "
+            "the DoFHandler passed to the constructor."));
+
+        const typename DoFHandler<dim>::active_cell_iterator dof_handler_cell(
+          &dof_handler->get_triangulation(),
+          cell->level(),
+          cell->index(),
+          dof_handler);
+
+        // Save the element and the local dof values, since this is what we need
+        // to evaluate the function.
+        element = &dof_handler_cell->get_fe();
+
+        local_dof_indices.resize(element->dofs_per_cell);
+        dof_handler_cell->get_dof_indices(local_dof_indices);
+
+        local_dof_values.resize(element->dofs_per_cell);
+
+        for (unsigned int i = 0; i < local_dof_indices.size(); i++)
+          local_dof_values[i] =
+            dealii::internal::ElementAccess<VectorType>::get(
+              *global_dof_values, local_dof_indices[i]);
+      }
+
+
+
+      template <int dim, class VectorType>
+      void
+      RefSpaceFEFieldFunction<dim, VectorType>::assert_cell_is_set() const
+      {
+        // If set cell hasn't been called the size of local_dof_values will be
+        // zero.
+        Assert(
+          local_dof_values.size() > 0,
+          ExcMessage(
+            "The set_active_cell function has to be called before calling this function."))
+      }
```

Comment:
We normally prefer if functions return bool and the calling function is asserting. You could even create an own excaption.